Main.hs

```diff
@@ -0,0 +1,149 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Main where
+
+import System.Environment
+import Text.ParserCombinators.Parsec hiding (spaces, try)
+import Text.Parsec.Prim
+import Control.Monad
+import Numeric
+import Data.Foldable
+import Data.Char
+
+symbol :: Parser Char
+symbol = oneOf "!#$%&|*+-/:<=>?@^_~"
+
+readExpr :: String -> String
+readExpr input = case parse parseExpr "scheme" input of
+  Left err -> "No match: " ++ show err
+  Right val -> "Found value:\n" ++ show val
+
+spaces :: Parser ()
+spaces = skipMany1 space
+
+main :: IO ()
+main = do
+  (expr:_) <- getArgs
+  putStrLn (readExpr expr)
+
+data LispValue = Atom String
+               | List [LispValue]
+               | DottedList [LispValue] LispValue
+               | RealNumber LispReal
+               | RationalNumber LispRational
+               | ComplexNumber LispComplex
+               | Character String
+               | String String
+               | Bool Bool
+               deriving (Eq, Show)
+
+data LispComplex =
+  LispComplex LispReal LispReal
+  deriving (Eq, Show)
+
+data LispReal =
+  LispDouble Double | LispRat LispRational | LispInteger Integer
+  deriving (Eq, Show)
+
+data LispRational = LispRational Integer Integer
+  deriving (Eq, Show)
+
+parseString :: Parser LispValue
+parseString = do
+  char '"'
+  x <- many1 ((char '\\' >> char '"')
```

Comment:
This is wrong – if you have e.g. `\r` in your string, it'll return `r`, but it should return `'\r'` instead.

``` haskell
x <- many1 $ choice [
  string "\\\"" $> '"',
  string "\\r" $> '\r',
  string "\\\\" $> '\\',
  ...
  ]
```